### ISSUE TITLE
How is setup/updater done?

### ISSUE THREAD

**Issue by urza** [canton7/SyncTrayzor#176]
Hi, I just wanted to ask, what setup tool is used to create the installer and updater? The update functionality is very good (I mean when there is new version of synctrayzor, that it will notify me, and if I agree, automatically update and restart itself). How is this done? It might help some other opensource project that could also benefit from such smooth updater.
Thanks

(I wanted to ask this on forums, but they have some error at the moment)

**Comment by canton7** [comment]
The installer's built using InnoSetup. SyncTrayzor periodically checks to see if there's a new version (which is appropriate for its current version, variant, and architecture, [see here](https://github.com/canton7/SyncTrayzor/blob/master/server/version_check.php)). If there is, it's downloaded and verified ([see here](https://github.com/canton7/SyncTrayzor/tree/master/security#automatic-update-security)), and a prompt displayed to the user.

If the user accepts the prompt, then the installer is launched (silently, and as admin, which takes a bit of effort). The installer process is detached from its parent, then Inno Setup automatically takes care of closing SyncTrayzor. Once the installation is complete, SyncTrayzor is restarted, and a notice displayed to the user.

It's a surprisingly involved process and took a good few weekends to write! Pretty much everything that's relevant is [in this folder](https://github.com/canton7/SyncTrayzor/tree/master/src/SyncTrayzor/Services/UpdateManagement), with [UpdateManager.cs](https://github.com/canton7/SyncTrayzor/blob/master/src/SyncTrayzor/Services/UpdateManagement/UpdateManager.cs) being the central point.

That's the top-level overview. Feel free to pester me further about particular bits you may have questions about.

**Comment by urza** [comment]
Thanks for the answer. It would be really cool to have the whole process and all steps in some tutorial, article, or something... I can imagine, getting it to work so nice and smooth requieres a lot of effort as you write, so to save time for other developers...

I will try to apply what you did to one of my c# wpf app and see if I can replicate that..

Is the setup signed with some paid certificate to avoid the windows warnings etc? And how exactly do you launch the updater as admin?

**Comment by canton7** [comment]
Yeah, good idea. I'll try and do a write-up when I have some time. Might be worth trying to split it into a library or something if and when I need it on another project.

The installer is signed, but only for legacy reasons: I used to verify the integrity of the downloaded installer by verifying its signature, but I now use a separate sha1sums file. 

Starting the installer as an elevated, detached process is a bitch. If you let the installer elevate itself, it will lose its command-line arguments. If you launch the installer elevated, then it will re-start SyncTrayzor as the admin, which is not what you want. 

I've got a separate project, [ProcessRunner](https://github.com/canton7/SyncTrayzor/tree/master/src/ProcessRunner), which basically just starts other processes (optionally as administrator, and either dying straight away (to orphan its child), or sticking around to launch another executable once the first one finishes).

In this case,  it's invoked by [`ProcessStartProvider.StartElevatedDetached`](https://github.com/canton7/SyncTrayzor/blob/master/src/SyncTrayzor/Services/ProcessStartProvider.cs#L64), which is in turn invoked from [`InstalledUpdateVariableHandler.AutoInstall`](https://github.com/canton7/SyncTrayzor/blob/master/src/SyncTrayzor/Services/UpdateManagement/InstalledUpdateVariantHandler.cs#L60).
